File: src/lib/hydrate-query.ts
import type {
	FilterGroup,
	FilterNode,
	FilterSpec,
	FilterValue,
	GroupConnector,
	JoinSpec,
	QuerySpec,
	TableInfo
} from '$lib/ir';
import { literalKindForColumn } from '$lib/ir';

export interface QueryHydrationTarget {
	clear(): void;
	addTable(tableInfo: TableInfo): void;
	readonly tables: ReadonlyArray<{ alias: string }>;
	toggleColumn(alias: string, column: string): void;
	addJoin(join: JoinSpec): void;
	setFilters(group: FilterGroup): void;
	setConnectorOverrides(map: Record<string, GroupConnector>): void;
	setLimit(limit: number): void;
}

export interface HydrationWarnings {
	droppedTables: string[];
	droppedJoins: number;
	droppedFilters: boolean;
}

function schemaKey(schema: string, name: string): string {
	return `${schema}\0${name}`;
}

function countFilterLeaves(group: FilterGroup): number {
	return group.children.reduce((sum, child) => {
		if ('Leaf' in child) return sum + 1;
		return sum + countFilterLeaves(child.Group);
	}, 0);
}

function normalizeFilterValue(value: FilterValue | null, dataType: string): FilterValue | null {
	if (!value) return null;
	const kind = literalKindForColumn(dataType);
	if ('Single' in value) {
		return { Single: { ...value.Single, kind } };
	}
	if ('List' in value) {
		return { List: value.List.map((literal) => ({ ...literal, kind })) };
	}
	return {
		Pair: [
			{ ...value.Pair[0], kind },
			{ ...value.Pair[1], kind }
		]
	};
}

function remapFilterGroup(
	group: FilterGroup,
	aliasMap: Map<string, string>,
	tableByNewAlias: Map<string, TableInfo>
): FilterGroup {
	const children: FilterNode[] = [];
	for (const child of group.children) {
		if ('Leaf' in child) {
			const tableAlias = aliasMap.get(child.Leaf.table_alias);
			const tableInfo = tableAlias ? tableByNewAlias.get(tableAlias) : null;
			const columnInfo = tableInfo?.columns.find((column) => column.name === child.Leaf.column);
			if (tableAlias && columnInfo) {
				children.push({
					Leaf: {
						...child.Leaf,
						table_alias: tableAlias,
						value: normalizeFilterValue(child.Leaf.value, columnInfo.data_type)
					}
				});
			}
		} else {
			const remapped = remapFilterGroup(child.Group, aliasMap, tableByNewAlias);
			if (remapped.children.length > 0) {
				children.push({ Group: remapped });
			}
		}
	}
	return { ...group, children };
}

/** Restore a saved or history query spec into the query store, remapping table aliases. */
export function hydrateQueryFromSpec(
	spec: QuerySpec,
	schemaTables: TableInfo[],
	target: QueryHydrationTarget
): HydrationWarnings {
	target.clear();

	const schemaByKey = new Map(
		schemaTables.map((table) => [schemaKey(table.schema, table.name), table])
	);
	const aliasMap = new Map<string, string>();
	const tableByNewAlias = new Map<string, TableInfo>();
	const droppedTables: string[] = [];

	for (const t of spec.tables) {
		const tableInfo = schemaByKey.get(schemaKey(t.schema, t.name));
		if (!tableInfo) {
			droppedTables.push(`${t.schema}.${t.name}`);
			continue;
		}

		target.addTable(tableInfo);
		const newAlias = target.tables[target.tables.length - 1]?.alias;
		if (newAlias) {
			aliasMap.set(t.alias, newAlias);
			tableByNewAlias.set(newAlias, tableInfo);
		}
	}

	for (const col of spec.columns) {
		const newAlias = aliasMap.get(col.table_alias);
		if (newAlias) {
			target.toggleColumn(newAlias, col.column);
		}
	}

	let droppedJoins = 0;
	for (const join of spec.joins) {
		const leftAlias = aliasMap.get(join.left_alias);
		const rightAlias = aliasMap.get(join.right_alias);
		if (!leftAlias || !rightAlias) {
			droppedJoins += 1;
			continue;
		}

		target.addJoin({
			left_alias: leftAlias,
			left_column: join.left_column,
			right_alias: rightAlias,
			right_column: join.right_column
		});
	}

	const originalFilterLeaves = countFilterLeaves(spec.filters);
	const remappedFilters = remapFilterGroup(spec.filters, aliasMap, tableByNewAlias);
	const droppedFilters = countFilterLeaves(remappedFilters) < originalFilterLeaves;
	target.setFilters(remappedFilters);

	target.setConnectorOverrides(spec.connector_overrides ?? {});

	target.setLimit(spec.limit);

	return { droppedTables, droppedJoins, droppedFilters };
}

export function formatHydrationWarning(warnings: HydrationWarnings): string | null {
	const parts: string[] = [];
	if (warnings.droppedTables.length > 0) {
		parts.push(`missing tables: ${warnings.droppedTables.join(', ')}`);
	}
	if (warnings.droppedJoins > 0) {
		parts.push(
			`${warnings.droppedJoins} join${warnings.droppedJoins !== 1 ? 's' : ''} could not be restored`
		);
	}
	if (warnings.droppedFilters) {
		parts.push('some filters were dropped');
	}
	if (parts.length === 0) return null;
	return `Query restored partially (${parts.join('; ')}). Review before running.`;
}

File: shared/src/main/kotlin/com/safedb/query/HydrateQuery.kt
package com.safedb.query

import com.safedb.model.FilterGroup
import com.safedb.model.FilterNode
import com.safedb.model.FilterValue
import com.safedb.model.GroupConnector
import com.safedb.model.GroupSpec
import com.safedb.model.JoinSpec
import com.safedb.model.QuerySpec
import com.safedb.model.SortSpec
import com.safedb.model.TableInfo

interface QueryHydrationTarget {
    fun clear()

    fun addTable(tableInfo: TableInfo)

    val tables: List<AliasRef>

    fun toggleColumn(alias: String, column: String)

    fun addJoin(join: JoinSpec)

    fun setFilters(group: FilterGroup)

    fun setConnectorOverrides(map: Map<String, GroupConnector>)

    fun setLimit(limit: Int)

    fun setDistinct(distinct: Boolean)

    fun setSorts(sorts: List<SortSpec>)

    fun setGroups(groups: List<GroupSpec>)
}

data class AliasRef(val alias: String)

data class HydrationWarnings(
    val droppedTables: List<String> = emptyList(),
    val droppedColumns: List<String> = emptyList(),
    val droppedJoins: Int = 0,
    val droppedFilters: Boolean = false,
    val droppedSorts: Int = 0,
    val droppedGroups: Int = 0,
)

private fun schemaKey(schema: String, name: String): String = "$schema\u0000$name"

fun countFilterLeaves(group: FilterGroup): Int =
    group.children.sumOf { child ->
        when (child) {
            is FilterNode.Leaf -> 1
            is FilterNode.Group -> countFilterLeaves(child.group)
        }
    }

private fun normalizeFilterValue(value: FilterValue?, dataType: String): FilterValue? {
    if (value == null) return null
    val kind = literalKindForColumn(dataType)
    return when (value) {
        is FilterValue.Single -> FilterValue.Single(value.literal.copy(kind = kind))
        is FilterValue.ListValue ->
            FilterValue.ListValue(value.literals.map { it.copy(kind = kind) })
        is FilterValue.Pair ->
            FilterValue.Pair(value.first.copy(kind = kind), value.second.copy(kind = kind))
    }
}

private fun remapFilterGroup(
    group: FilterGroup,
    aliasMap: Map<String, String>,
    tableByNewAlias: Map<String, TableInfo>,
): FilterGroup {
    val children = mutableListOf<FilterNode>()
    for (child in group.children) {
        when (child) {
            is FilterNode.Leaf -> {
                val tableAlias = aliasMap[child.spec.tableAlias]
                val tableInfo = tableAlias?.let { tableByNewAlias[it] }
                val columnInfo = tableInfo?.columns?.find { it.name == child.spec.column }
                if (tableAlias != null && columnInfo != null) {
                    children.add(
                        FilterNode.Leaf(
                            child.spec.copy(
                                tableAlias = tableAlias,
                                value = normalizeFilterValue(child.spec.value, columnInfo.dataType),
                            )
                        )
                    )
                }
            }
            is FilterNode.Group -> {
                val remapped = remapFilterGroup(child.group, aliasMap, tableByNewAlias)
                if (remapped.children.isNotEmpty()) {
                    children.add(FilterNode.Group(remapped))
                }
            }
        }
    }
    return group.copy(children = children)
}

fun hydrateQueryFromSpec(
    spec: QuerySpec,
    schemaTables: List<TableInfo>,
    target: QueryHydrationTarget,
): HydrationWarnings {
    target.clear()

    val schemaByKey = schemaTables.associateBy { schemaKey(it.schema, it.name) }
    val aliasMap = mutableMapOf<String, String>()
    val tableByNewAlias = mutableMapOf<String, TableInfo>()
    val droppedTables = mutableListOf<String>()
    val droppedColumns = mutableListOf<String>()

    for ((tableSchema, tableName, alias) in spec.tables) {
        val tableInfo = schemaByKey[schemaKey(tableSchema, tableName)]
        if (tableInfo == null) {
            droppedTables.add("$tableSchema.$tableName")
            continue
        }

        target.addTable(tableInfo)
        val newAlias = target.tables.lastOrNull()?.alias
        if (newAlias != null) {
            aliasMap[alias] = newAlias
            tableByNewAlias[newAlias] = tableInfo
        }
    }

    for ((tableAlias, column) in spec.columns) {
        val newAlias = aliasMap[tableAlias]
        val tableInfo = newAlias?.let { tableByNewAlias[it] }
        if (newAlias != null && tableInfo?.columns?.any { it.name == column } == true) {
            target.toggleColumn(newAlias, column)
        } else {
            droppedColumns.add("$tableAlias.$column")
        }
    }

    var droppedJoins = 0
    for ((joinLeftAlias, leftColumn, joinRightAlias, rightColumn) in spec.joins) {
        val leftAlias = aliasMap[joinLeftAlias]
        val rightAlias = aliasMap[joinRightAlias]
        val leftTable = leftAlias?.let { tableByNewAlias[it] }
        val rightTable = rightAlias?.let { tableByNewAlias[it] }
        val leftColumnExists = leftTable?.columns?.any { it.name == leftColumn } == true
        val rightColumnExists = rightTable?.columns?.any { it.name == rightColumn } == true
        if (leftAlias == null || rightAlias == null || !leftColumnExists || !rightColumnExists) {
            droppedJoins += 1
            continue
        }

        target.addJoin(
            JoinSpec(
                leftAlias = leftAlias,
                leftColumn = leftColumn,
                rightAlias = rightAlias,
                rightColumn = rightColumn,
            )
        )
    }

    val originalFilterLeaves = countFilterLeaves(spec.filters)
    val remappedFilters = remapFilterGroup(spec.filters, aliasMap, tableByNewAlias)
    val droppedFilters = countFilterLeaves(remappedFilters) < originalFilterLeaves
    target.setFilters(remappedFilters)

    target.setConnectorOverrides(spec.connectorOverrides)
    target.setLimit(spec.limit)
    target.setDistinct(spec.distinct)

    var droppedSorts = 0
    target.setSorts(
        spec.sorts.mapNotNull { sort ->
            val newAlias = aliasMap[sort.tableAlias]
            val tableInfo = newAlias?.let { tableByNewAlias[it] }
            if (newAlias == null || tableInfo?.columns?.none { it.name == sort.column } != false) {
                droppedSorts += 1
                null
            } else {
                sort.copy(tableAlias = newAlias)
            }
        }
    )

    var droppedGroups = 0
    target.setGroups(
        spec.groups.mapNotNull { group ->
            val newAlias = aliasMap[group.tableAlias]
            val tableInfo = newAlias?.let { tableByNewAlias[it] }
            if (newAlias == null || tableInfo?.columns?.none { it.name == group.column } != false) {
                droppedGroups += 1
                null
            } else {
                group.copy(tableAlias = newAlias)
            }
        }
    )

    return HydrationWarnings(
        droppedTables = droppedTables,
        droppedColumns = droppedColumns,
        droppedJoins = droppedJoins,
        droppedFilters = droppedFilters,
        droppedSorts = droppedSorts,
        droppedGroups = droppedGroups,
    )
}

fun formatHydrationWarning(warnings: HydrationWarnings): String? {
    val parts = mutableListOf<String>()
    if (warnings.droppedTables.isNotEmpty()) {
        parts.add("missing tables: ${warnings.droppedTables.joinToString(", ")}")
    }
    if (warnings.droppedColumns.isNotEmpty()) {
        val count = warnings.droppedColumns.size
        parts.add("$count selected column${if (count != 1) "s" else ""} could not be restored")
    }
    if (warnings.droppedJoins > 0) {
        val count = warnings.droppedJoins
        parts.add("$count join${if (count != 1) "s" else ""} could not be restored")
    }
    if (warnings.droppedFilters) {
        parts.add("some filters were dropped")
    }
    if (warnings.droppedSorts > 0) {
        val count = warnings.droppedSorts
        parts.add("$count sort${if (count != 1) "s" else ""} could not be restored")
    }
    if (warnings.droppedGroups > 0) {
        val count = warnings.droppedGroups
        parts.add("$count group${if (count != 1) "s" else ""} could not be restored")
    }
    if (parts.isEmpty()) return null
    return "Query restored partially (${parts.joinToString("; ")}). Review before running."
}
